contractsSol5/KyberNetwork.sol

```diff
@@ -478,297 +470,225 @@ contract KyberNetwork is Withdrawable3, Utils4, IKyberNetwork, ReentrancyGuard {
+        destAmout = validateTradeCalcDestQtyAndFeeData(tradingReserves, tData);
     }
 
-    /// @notice extra processing, currently for non-splitting trade type
-    ///         calling reserves to get rates
-    ///         calling matching engine to process data with rates
-    ///         save list of chosen reserves for final check and process
-    function getTradeDataExtraProcessing(
-        TradingReserves memory tradingReserves,
-        IERC20 token,
-        uint srcAmount,
-        bool isTokenToEth,
-        uint networkFee, // bps in case t2e, wei value in case e2t
-        bytes8[] memory reserveIDs,
-        uint[] memory splitValuesBps,
-        bool[] memory isFeeCounted
-    )
+    /// @dev calculates source amounts per reserve. does get rate call.
+    function calcSrcAmountsAndGetRates(
+            TradingReserves memory tradingReserves,
+            IERC20 src,
+            uint srcAmount,
+            IERC20 dest,
+            uint networkFee
+        )
         internal view
     {
-        uint[] memory srcAmounts = new uint[](reserveIDs.length);
-        uint[] memory rates = new uint[](reserveIDs.length);
-        // this value is only used for t2e trade
-        uint[] memory feeAccountedBps;
-        if (isTokenToEth) {
-            feeAccountedBps = new uint[](reserveIDs.length);
-        }
-
-        for(uint i = 0; i < reserveIDs.length; i++) {
-            require(splitValuesBps[i] > 0 && splitValuesBps[i] <= BPS, "invalid split bps");
-            IKyberReserve reserve = IKyberReserve(convertReserveIdToAddress(reserveIDs[i]));
-            require(reserve != IKyberReserve(0), "reserve is not listed");
-
-            if (isTokenToEth) {
-                srcAmounts[i] = srcAmount * splitValuesBps[i] / BPS;
-                feeAccountedBps[i] = isFeeCounted[i] ? networkFee : 0;
-            } else {
-                require(srcAmount >= networkFee, "srcAmount is less than networkFee");
-                srcAmounts[i] = isFeeCounted[i] ? (srcAmount - networkFee) * splitValuesBps[i] / BPS :
-                    srcAmount * splitValuesBps[i] / BPS;
-            }
+        uint numReserves = tradingReserves.ids.length;
+
+        tradingReserves.srcAmounts = new uint[](numReserves);
+        tradingReserves.rates = new uint[](numReserves);
+        tradingReserves.feeAccountedBpsDest = new uint[](numReserves);
+        tradingReserves.addresses = new IKyberReserve[](numReserves);
+
+        // iterate reserve list. validate data. calculate srcAmount according to splits and fee data.
+        for (uint i = 0; i < numReserves; i++) {
+            require(tradingReserves.splitsBps[i] > 0 && tradingReserves.splitsBps[i] <= BPS, "invalid split bps");
+            tradingReserves.addresses[i] = IKyberReserve(convertReserveIdToAddress(tradingReserves.ids[i]));
+            
+            tradingReserves.srcAmounts[i] = srcAmount * tradingReserves.splitsBps[i] / BPS;
```

Comment:
This part of src amounts is not correct for splitting trade (e2t) compare to our current logic.
For t2e, it is correct that `srcAmounts[i] = srcAmount * splitsBps[i] / BPS`
For e2t, if it is splitting trade, we compute total `feePayingReserveBps` (combine of t2e and e2t), then compute total `networkFeeWei`. After that we have `actualSrcAmount` for e2t trade (`tradeWei - networkFeeWei - platformFeeWei`, and `srcAmounts[i] = actualSrcAmount * splitsBps[i] / BPS`.


Take an example:
assume: networkFee: 0.25%, platformFee = 0
t2e: done, tradeWei = 1 eth, networkFeeWei = 0.25% * 1 = 0.0025 eth (feePayingReserveBps = 100%)
e2t: split into 2 reserves, each 50%, one reserve is paying fee, one isn't.

Current logic:
- feePayingReserveBps = 100% (t2e) + 50% (e2t) = 150%
- networkFeeWei = 1 * 0.25% * 150% = 0.00375 eth
- actualSrcAmount = 1 - 0.00375 = 0.99625 eth
- src amount for each reserve in e2t trade: 0.99625 * 50% = 0.498125 eth

New logic:
- srcAmount for e2t: 1 - 0.25% * 1 = 0.9975 eth
- srcAmount each reserve: srcAmounts[i] = srcAmount * 50% = 0.49875 eth
- srcAmount reserve not paying fee = 0.49875 eth
- srcAmount reserve paying fee = 0.49875 * (100% - 0.25%) = 0.497503125 eth (in line 593)

Reply:
OK

Reply:
fixed
with a bit new logic
See here
https://github.com/KyberNetwork/smart-contracts/pull/807/commits/4e5194171f76759b69b7b4c19731d88ec28851aa

last section